File: Data_Structure/Weighted-UnionFind.py
# verified with:
# https://onlinejudge.u-aizu.ac.jp/courses/library/3/DSL/all/DSL_1_B
# https://atcoder.jp/contests/abc087/submissions/39746989
# https://atcoder.jp/contests/abc280/submissions/39747056

from typing import TypeVar, Callable, List

T = TypeVar("T")
class Weighted_UnionFind:
    def __init__(self, _n:int, _fx:Callable[[T,T],T], _inv:Callable[[T],T], _ex:T) -> None:
        self.n = _n
        self.fx = _fx
        self.inv = _inv
        self.ex = _ex
        self.parents = [-1]*_n
        self.potential = [_ex]*_n

    def merge(self, x:int, y:int, w:T) -> bool:
        w += self.weight(x)
        w -= self.weight(y)
        x = self.root(x)
        y = self.root(y)
        if x == y:
            return self.diff(x,y) == w
        else:
            if self.parents[x] > self.parents[y]:
                x,y = y,x
                w = self.inv(w)
            self.parents[x] += self.parents[y]
            self.parents[y] = x
            self.potential[y] = w
            return True

    def same(self, x:int, y:int) -> bool:
        return self.root(x) == self.root(y)
    
    def root(self, x:int) -> int:
        if self.parents[x] < 0:
            return x
        else:
            r = self.root(self.parents[x])
            self.potential[x] = self.fx(self.potential[x],self.potential[self.parents[x]])
            self.parents[x] = r
            return self.parents[x]
    
    def size(self, x:int) -> int:
        return -self.parents[self.root(x)]

    def weight(self, x:int) -> T:
        self.root(x)
        return self.potential[x]
    
    def diff(self, x:int, y:int) -> T:
        return self.fx(self.weight(y),self.inv(self.weight(x)))
  
    def groups(self) -> List[List[int]]:
        ret = [[] for i in range(self.n)]
        for i in range(self.n):
            ret[self.root(i)].append(i)
        return [i for i in ret if len(i)]

File: Data_Structure/Weighted-UnionFind.cpp
// verified with:
// https://onlinejudge.u-aizu.ac.jp/courses/library/3/DSL/all/DSL_1_B
// https://atcoder.jp/contests/abc087/submissions/39744954
// https://atcoder.jp/contests/abc280/submissions/39745774

#include<bits/stdc++.h>
using namespace std;


template<typename T>
struct Weighted_UnionFind{
  Weighted_UnionFind() : n(0) {}
  explicit Weighted_UnionFind(int n, function<T(T,T)> _fx, function<T(T)> _inv, T _ex) : n(n), fx(_fx), inv(_inv){
    parents.assign(n,-1);
    potential.assign(n,_ex);
  }
  
  bool merge(int x, int y, T w){
    w += weight(x); w -= weight(y);
    x = root(x), y = root(y);

    // 既に同じ集合に属しているときの処理
    if (x == y) {
      // 矛盾が無ければtrue、あればfalse
      if (diff(x,y) == w){
        return true;
      }else{
        return false;
      }
    }

    if (parents[x] > parents[y]) swap(x,y), w = inv(w);
    parents[x] += parents[y];
    parents[y] = x;
    potential[y] = w;
		return true;
  }

  bool same(int x, int y){
    return root(x) == root(y);
  }

  int root(int x){
    if (parents[x] < 0){
      return x;
    }else{
      int r = root(parents[x]);
      potential[x] = fx(potential[x],potential[parents[x]]);
      return parents[x] = r;
    }
  }

  int size(int x){
    return -parents[root(x)];
  }

  T weight(int x){
    root(x);
    return potential[x];
  }

  T diff(int x, int y) {
    return fx(weight(y),inv(weight(x)));
  }
  
  vector<vector<int>> groups(){
    vector<vector<int>> ret(n);
    vector<int> cnt(n),roots(n);
    for(int i = 0; i < n; ++i) {
      roots[i] = root(i);
      ++cnt[roots[i]];
    }
    for(int i = 0; i < n; ++i) ret[i].reserve(cnt[i]);
    for(int i = 0; i < n; ++i) ret[roots[i]].push_back(i);
    ret.erase(remove_if(ret.begin(), ret.end(), [&](const std::vector<int>& i) {return i.empty();}), ret.end());
    return ret;
  }

  private:
    const int n;
    vector<int> parents;
    vector<T> potential;
		function<T(T,T)> fx;
		function<T(T)> inv;
};